swift/extractor/SwiftOutputRewrite.cpp

```diff
@@ -0,0 +1,329 @@
+  secondCase += moduleName.str() + '/';
+  secondCase += moduleNameWithExt.str() + '/';
+  moduleLocations.push_back(secondCase);
+
+  std::string thirdCase = relocatedModulePath;
+  thirdCase += moduleName.str() + '/';
+  thirdCase += moduleName.str() + ".framework/Modules/";
+  thirdCase += moduleNameWithExt.str() + '/';
+  moduleLocations.push_back(thirdCase);
+
+  std::vector<std::string> aliases;
+  for (auto& location : moduleLocations) {
+    aliases.push_back(location + arch + ".swiftmodule");
+    if (!targetTriple.empty()) {
+      llvm::Triple triple(targetTriple);
+      auto moduleTriple = swift::getTargetSpecificModuleTriple(triple);
+      aliases.push_back(location + moduleTriple.normalize() + ".swiftmodule");
+    }
+  }
+
+  return aliases;
+}
+
+namespace codeql {
+
+std::unordered_map<std::string, std::string> rewriteOutputsInPlace(
+    SwiftExtractorConfiguration& config,
+    std::vector<std::string>& CLIArgs) {
+  std::unordered_map<std::string, std::string> remapping;
+
+  // TODO: handle filelists?
+  std::unordered_set<std::string> pathRewriteOptions({
+      "-emit-dependencies-path",
+      "-emit-module-path",
+      "-emit-module-doc-path",
+      "-emit-module-source-info-path",
+      "-emit-objc-header-path",
+      "-emit-reference-dependencies-path",
+      "-index-store-path",
+      "-module-cache-path",
+      "-o",
+      "-pch-output-dir",
+      "-serialize-diagnostics-path",
+  });
+
+  std::unordered_set<std::string> outputFileMaps(
+      {"-supplementary-output-file-map", "-output-file-map"});
+
+  std::vector<size_t> outputFileMapIndexes;
+  std::vector<std::string> maybeInput;
+  std::string targetTriple;
+
+  std::vector<std::string> newLocations;
+  for (size_t i = 0; i < CLIArgs.size(); i++) {
+    if (pathRewriteOptions.count(CLIArgs[i])) {
+      auto oldPath = CLIArgs[i + 1];
+      auto newPath = config.tempArtifactDir + '/' + oldPath;
+      CLIArgs[++i] = newPath;
+      newLocations.push_back(newPath);
+
+      remapping[oldPath] = newPath;
+    } else if (outputFileMaps.count(CLIArgs[i])) {
+      // collect output map indexes for further rewriting and skip the following argument
+      // We don't patch the map in place as we need to collect all the input files first
+      outputFileMapIndexes.push_back(++i);
+    } else if (CLIArgs[i] == "-target") {
+      targetTriple = CLIArgs[++i];
+    } else if (CLIArgs[i][0] != '-') {
+      // TODO: add support for input file lists?
+      // We need to collect input file names to later use them to extract information from the
+      // output file maps.
+      maybeInput.push_back(CLIArgs[i]);
+    }
+  }
+
+  for (auto index : outputFileMapIndexes) {
+    auto oldPath = CLIArgs[index];
+    auto maybeNewPath = rewriteOutputFileMap(config, oldPath, maybeInput, remapping);
+    if (maybeNewPath) {
+      auto newPath = maybeNewPath.value();
+      CLIArgs[index] = newPath;
+      remapping[oldPath] = newPath;
+    }
+  }
```

Comment:
there are unneeded string copies
```suggestion
  for (auto index : outputFileMapIndexes) {
    auto maybeNewPath = rewriteOutputFileMap(config, CLIArgs[index], maybeInput, remapping);
    if (maybeNewPath) {
      oldPath = std::move(CLIArgs[index]);
      CLIArgs[index] = *maybeNewPath;
      remapping.insert_or_assign(std::move(oldPath), std::move(newPath));
    }
  }
```

Reply:
There is an ongoing pattern in this file with `remapping[oldPath] = newPath`. I'd like to stick to it (even in cases where it's not the most efficient way) just for the sake of consistency.

Reply:
well, I would consistently use `insert_or_assign`, which I suggested everywhere :smile:

Reply:
(but not a blocker for me, just to be clear)